### DISCUSSION TITLE
Customizing complete menu

### DISCUSSION THREAD

**Discussion by romaia** [akinomyoga/ble.sh#433] [Q&A] [answered]
Is it possible to customize the menu for the tab completions?

For instance. This is how completion looks in fish:

![image](https://github.com/akinomyoga/ble.sh/assets/428053/a41e0d21-5f8f-44aa-a171-15a0be4ad5bb)

And this is the current default in ble.sh

![image](https://github.com/akinomyoga/ble.sh/assets/428053/6d7b0f15-e418-4389-bca1-2c9f6a2ad8e7)

I would like to use a ligher color for the description, to focus on the commands itself.

**Comment by akinomyoga** [comment]
> I would like to use a ligher color for the description, to focus on the commands itself.

I'm not sure if I got it correctly, but would you like to specify the graphic style of the description of the selected item?

**Reply by akinomyoga** [reply]
OK, I now understand what you request.

I thought you tried to show the description of the selected item in the reverse video as fish does, but then I was wondering why it makes you focus on the commands themselves because it seems to drive more attention to the description of the selected item instead of the commands.

If you want to change the color of the descriptions in general, that makes sense. There is no option to directly specify it, but you can include the ANSI control sequences in the description (such as the SGR codes of `$'\e[33mdescription...\e[m'` with `bleopt complete_menu_style=desc`. An SGR code can be embedded in the description at the generation phase. The completion settings built in ble.sh don't provide the completions for `mgc`, so I guess the completion is generated by some other framework or the settings you made. Where did you get the completion setting of `mgc`?

**Reply by akinomyoga** [reply]
OK, I think `mgc` uses the completion framework `cobra`, which is actually supported by ble.sh. I couldn't find anything on the internet about the command `mgc`, but the description ["Generate the autocompletion script for the specified shell"](https://github.com/spf13/cobra/blob/0fc86c2ffd0326b6f6ed5fa36803d26993655c08/completions.go#L706) is specific to `cobra`. With the current implementation of ble.sh, the description from cobra is not highlighted because `cobra` doesn't include SGR codes to highlight the description, and ble.sh just uses the description generated by `cobra` as is.

https://github.com/akinomyoga/ble.sh/blob/adf9ab8197ec3e7d52875e23374491cfd498f524/lib/core-complete.sh#L3796-L3801



To allow configuring the color of the description, I think it is easier to add a global option in ble.sh to specify the *default color* of the description.

**Reply by romaia** [reply]
mgc is the cli of a new cloud being launched in brasil (magalu.cloud). Source code is not available yet, but it can be downloaded here: https://github.com/MagaluCloud/mgccli (disclaimer: I work at magalu cloud - but not the on the cli)

Yes, it does use cobra.

> but would you like to specify the graphic style of the description of the selected item

Yes!

>  it seems to drive more attention to the description of the selected item instead of the commands.

I agree, but on the bash with ble.sh (first screenshot), since both the command and the description use the same color, It makes it harder to see the commands, that why I would like to use a less proeminent color on it.

> and ble.sh just uses the description generated by cobra as is.

Do you think this is something worth supporting? I can work around this editing the description for now, but this seems like a nice addition.

**Reply by akinomyoga** [reply, marked as answer]
In commit b3e7237e31fa00f347ca7c8227f44f37df85f059, I added the following new faces:

- [`ble-face menu_desc_default`](https://github.com/akinomyoga/ble.sh/wiki/Manual-%C2%A77-Completion#user-content-face-menu_desc_default)
- [`ble-face menu_desc_type`](https://github.com/akinomyoga/ble.sh/wiki/Manual-%C2%A77-Completion#user-content-face-menu_desc_type)
- [`ble-face menu_desc_quote`](https://github.com/akinomyoga/ble.sh/wiki/Manual-%C2%A77-Completion#user-content-face-menu_desc_quote)

You can now change the default color of the menu description by `ble-face menu_desc_default`. For example, you can use the following setting:

```bash
ble-face menu_desc_default='fg=olive'
```

@romaia Could you update ble.sh by running `ble-update` and check the behavior?

**Reply by romaia** [reply]
\o/

It works perfecly. Thankyou very much.

While we are talking about menu customizations, have you ever considered one that looks like a popup? something in the way neovim does?

Inshellisense does something that also looks nice (https://github.com/microsoft/inshellisense/)

**Reply by akinomyoga** [reply]
Thanks for checking the behavior.

For the popup completion menu, I did consider it as a possible option before I started to implement the completion menu, but I gave up implementing it. Some considerations and discussions can be found in the following links (though they are in Japanese):

- https://github.com/akinomyoga/ble.sh/issues/5#issuecomment-423821900
- https://github.com/akinomyoga/ble.sh/blob/56765251c3d7818d044f2703e121c55817c72986/note.txt#L6993-L7014
- https://github.com/akinomyoga/ble.sh/blob/56765251c3d7818d044f2703e121c55817c72986/note.txt#L6416-L6443

**Reply by akinomyoga** [reply]
There are two reasons that I don't feel it would be worth implementing.

- The number of items and the amount of information shown at once are limited with the popup-window interface. It looks fancy, but I'm not sure if I would continue to use it. Even if I would implement it, after enjoying it for a few hours, I'd probably go back to the full-width completion menu and will never turn on it again.
- This adds unnecessary complication to the implementation. To show a popup window, the program needs to remember the rendered content hidden by the popup window so that it can rerender the content after the popup window changes its size or position or is closed. Since ble.sh offers multiline editing, we cannot simply clear the terminal region that the popup window has occupied. If it is properly implemented in a shell script, it would also have a performance issue because the shell language does not provide the proper data structures to implement it. We need to implement everything using strings and one-dimensional string arrays. As another way, we could rerender the entire prompt and the command line every time the popup window changes its status, but this will add excessive flickering.

I wouldn't implement a feature just for the reason that it "looks nice". Do you think there are practical benefits in having the popup windows?

**Reply by romaia** [reply]
Nope, thats just something I was wondering. Although it looks nice, I agree the benefits are not worth the additional complexity.
